feat: Implement product search and insertion

ananiasc opened the pull request.

ananias-ifce committed:
feat: Implement product search and insertion

ronierlima commented on pom.xml:
Regarding the feature/add-product-search, it includes much more than it should, such as adding several packages in pom.xml, like Lombok, Hibernate, and so on. If I reject your feature, would I not have these packages?

ronierlima replied:
Would I create features for each new package addition and configuration, or one specifically to add those that will be used. ex.: feature/project-start

ronierlima commented on src/main/java/br/com/ananiascaetano/application/services/product/ProductService.java:
```diff
@@ -0,0 +1,34 @@
+	@Autowired
+	ProductRepository repository;
+	
+	@Autowired
```
I do not recommend using @Autowired. Besides being deprecated, you may encounter problems. [Link to article](https://www.baeldung.com/java-spring-field-injection-cons).

ronierlima replied:
I recommend using ```private final ProductRepository repository```; and creating a constructor. If you do not want to use the constructor, you can use @RequiredArgsConstructor. [Lombok Constructor](https://projectlombok.org/features/constructor), [Baeldung Lombok Constructor](https://www.baeldung.com/java-lombok-constructor-annotation).

ronierlima commented on src/main/java/br/com/ananiascaetano/application/services/product/ProductService.java:
```diff
@@ -0,0 +1,34 @@
+	}
+	
+	public Product createProduct(ProductDTO productDTO) {
+		ProductType type = typeRepository.findByName(productDTO.getType())
```
I would create a service for ProductType and place the search logic there.

ronierlima commented on src/main/java/br/com/ananiascaetano/presentation/controllers/product/ProductController.java:
```diff
@@ -0,0 +1,27 @@
+import br.com.ananiascaetano.presentation.dtos.product.ProductDTO;
+
+@RestController
+@RequestMapping("/portal/products")
```
What is /portal?

ronierlima commented on src/main/java/br/com/ananiascaetano/application/services/product/ProductService.java:
```diff
@@ -0,0 +1,34 @@
+	
+	public Product createProduct(ProductDTO productDTO) {
+		ProductType type = typeRepository.findByName(productDTO.getType())
+				.orElseThrow(() ->  new RuntimeException("Product type not found"));
```
I would put these messages in a ``` public final static String MESSAGE ```. Even better if a file is created for this in the future.

ronierlima commented on src/main/java/br/com/ananiascaetano/domain/entities/product/Product.java:
```diff
@@ -0,0 +1,49 @@
+import lombok.NoArgsConstructor;
+
+@Data
+@Builder
```
I would only use `@Builder` if I were really going to use the builders.

ronierlima commented on src/main/java/br/com/ananiascaetano/presentation/controllers/product/ProductController.java:
```diff
@@ -0,0 +1,27 @@
+
+	@GetMapping
+	public List<ProductDTO> findAll() {
+		return productService.findAll().stream().map(obj -> new ProductDTO(obj)).collect(Collectors.toList());
```
I suggest using the modelMapper package [Baeldung link](https://www.baeldung.com/java-modelmapper) for DTO parsing.

ronierlima commented on src/main/java/br/com/ananiascaetano/presentation/dtos/product/ProductDTO.java:
```diff
@@ -0,0 +1,75 @@
+	}
+	
+	@JsonIgnore
+	public boolean isDiscounted() {
```
If you're going to ignore it, why is it in the DTO?

ronierlima commented on src/main/java/br/com/ananiascaetano/presentation/dtos/product/ProductDTO.java:
```diff
@@ -0,0 +1,75 @@
+	}
+	
+	@JsonIgnore
+	public boolean isActive() {
```
If you're going to ignore it, why is it in the DTO?

ananias-ifce committed:
fix: Overriding deprecated @Autowired

Replacing deprecated @Autowired with @RequiredArgsConstructor for dependency injection of objects defined with private final

ananias-ifce committed:
feat: Adding library for entity mapping

ananias-ifce committed:
feat: performing mapping from ModelMapper to product list

ananias-ifce committed:
fix: performing mapping from ModelMapper to save product and return its DTO

ananias-ifce committed:
fix: remove findByName of ProductTypeRepository

ananias-ifce committed:
fix: removing unused functions

ananias-ifce committed:
fix: changing variable names

ananias-ifce committed:
feat: creating BaseDTO and using it as a generic for objects with only id

ananias-ifce committed:
fix: removing unused @Builder

ananias-ifce committed:
fix: separating search responsibility from product type to specific service

ananias-ifce committed:
fix: removing portal from RequestMapping

ananias-ifce committed:
fix: transferring data mapping responsibility to the controller

ananias-ifce committed:
refactor: creating class for error message responsibility

ananias-ifce committed:
feat: creating RuntimeExpection extension with message

ananias-ifce committed:
feat: creating ControlerAdvice for handling known errors

ananias-ifce committed:
feat: applying use of created Exception
